src/backend/distributed/connection/connection_management.c

```diff
@@ -438,183 +466,372 @@ ShutdownConnection(MultiConnection *connection)
 
 
 /*
- * FinishConnectionListEstablishment is a wrapper around FinishConnectionEstablishment.
- * The function iterates over the multiConnectionList and finishes the connection
- * establishment for each multi connection.
+ * MultiConnectionStatePoll executes a PQconnectPoll on the connection to progres the
+ * connection establishment. The return value of this function indicates if the WaitSet
+ * should be updated (when the poll mode changes), rebuilt (when no more polling for this
+ * connections is required) or Nothing (when the pollmode stays the same).
  */
-void
-FinishConnectionListEstablishment(List *multiConnectionList)
+static enum EventSetResult
+MultiConnectionStatePoll(MultiConnectionState *connectionState)
 {
-	ListCell *multiConnectionCell = NULL;
+	MultiConnection *connection = connectionState->connection;
+	ConnStatusType status = PQstatus(connection->pgConn);
+	PostgresPollingStatusType oldPollmode = connectionState->pollmode;
 
-	foreach(multiConnectionCell, multiConnectionList)
+	Assert(connectionState->phase == MULTI_CONNECTION_PHASE_CONNECTING);
+
+	if (status == CONNECTION_OK)
+	{
+		connectionState->phase = MULTI_CONNECTION_PHASE_CONNECTED;
+		return REBUILD_WAITEVENTSET;
+	}
+	else if (status == CONNECTION_BAD)
 	{
-		MultiConnection *multiConnection = (MultiConnection *) lfirst(
-			multiConnectionCell);
+		/* FIXME: retries? */
+		connectionState->phase = MULTI_CONNECTION_PHASE_ERROR;
+		return REBUILD_WAITEVENTSET;
+	}
+	else
+	{
+		connectionState->phase = MULTI_CONNECTION_PHASE_CONNECTING;
+	}
 
-		/* TODO: consider making connection establishment fully in parallel */
-		FinishConnectionEstablishment(multiConnection);
+	connectionState->pollmode = PQconnectPoll(connection->pgConn);
+
+	/*
+	 * FIXME: Do we want to add transparent retry support here?
+	 */
+	if (connectionState->pollmode == PGRES_POLLING_FAILED)
+	{
+		connectionState->phase = MULTI_CONNECTION_PHASE_ERROR;
+		return REBUILD_WAITEVENTSET;
+	}
+	else if (connectionState->pollmode == PGRES_POLLING_OK)
+	{
+		connectionState->phase = MULTI_CONNECTION_PHASE_CONNECTED;
+		return REBUILD_WAITEVENTSET;
 	}
+	else
+	{
+		Assert(connectionState->pollmode == PGRES_POLLING_WRITING ||
+			   connectionState->pollmode == PGRES_POLLING_READING);
+	}
+
+	return (oldPollmode != connectionState->pollmode) ? UPDATE_EVENT_MASK : NONE;
 }
 
 
 /*
- * Synchronously finish connection establishment of an individual connection.
+ * eventSetSizeForConnectionList calculates the space needed for a WaitEventSet based on a
+ * list of connections.
+ */
+inline static int
+eventSetSizeForConnectionList(List *connections)
```

Comment:
I'm not sure this still adds much, it might be helpful to actually inline because the number of extra events depends on what the caller does.

Reply:
As discussed, we will keep it as a function (altough with a capitalised name) since it is used on 2 occasions.

 - `WaitEventSetFromMultiConnectionStates`
 - `FinishConnectionListEstablishment`